pkg/serviceaccounttoken/secret_cleaner.go

```diff
@@ -0,0 +1,196 @@
+}
+
+type serviceAccountsCache interface {
+	GetByIndex(indexName, key string) ([]*corev1.ServiceAccount, error)
+	AddIndexer(indexName string, indexer generic.Indexer[*corev1.ServiceAccount])
+}
+
+// setupServiceAccountsCache must be called early on to setup the correct
+// indexing on service accounts.
+func setupServiceAccountsCache(cache serviceAccountsCache) {
+	cache.AddIndexer(serviceAccountSecretRefIndex, func(s *corev1.ServiceAccount) ([]string, error) {
+		if s.ObjectMeta.Annotations == nil {
+			return nil, nil
+		}
+		annotation := s.ObjectMeta.Annotations[ServiceAccountSecretRefAnnotation]
+		if annotation != "" {
+			return []string{annotation}, nil
+		}
+
+		return nil, nil
+	})
+}
+
+// StartServiceAccountSecretCleaner starts a background process to cleanup old
+// service accounts secrets.
+//
+// This should only be started in the leader pod.
+func StartServiceAccountSecretCleaner(ctx context.Context, secrets secretsCache, serviceAccounts serviceAccountsCache, client clientcorev1.CoreV1Interface) error {
+	if !features.CleanStaleSecrets.Enabled() {
+		logrus.Info("ServiceAccountSecretCleaner disabled - not starting")
+		return nil
+	}
+
+	setupServiceAccountsCache(serviceAccounts)
+
+	secretsQueue, err := loadSecretQueue(secrets)
+	if err != nil {
+		return err
+	}
+
+	startTime := time.Now()
+
+	logrus.Infof("Starting ServiceAccountSecretCleaner with %v secrets", secretsQueue.List.Len())
+	ticker := time.NewTicker(cleanCycleDelay)
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				logrus.Info("terminating service account secret cleaner")
+				return
+			case <-ticker.C:
+				if err := CleanServiceAccountSecrets(ctx, client.Secrets(impersonationNamespace), serviceAccounts, secretsQueue.dequeue(cleaningBatchSize)); err != nil {
+					logrus.Error(err)
+				}
+				if l := secretsQueue.List.Len(); l > 0 {
+					logrus.Infof("ServiceAccountSecretCleaner has %v secrets remaining", l)
+				} else {
+					logrus.Infof("ServiceAccountSecretCleaner has no secrets remaining - terminating at %v", time.Since(startTime))
+					return
+				}
+				// This ensures that no matter how long the cleaning takes,
+				// we'll alwas keep the same cycle time.
+				ticker.Reset(cleanCycleTime)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// CleanServiceAccountSecrets removes unused Secrets for ServiceAccountTokens from a
+// namespace in batches.
+func CleanServiceAccountSecrets(ctx context.Context, secrets clientv1.SecretInterface, serviceAccounts serviceAccountsCache, secretsToDelete []types.NamespacedName) error {
+	var deletionErr error
+	var deletedCount int64
+	var toBeDeleted []types.NamespacedName
+
+	for i := range secretsToDelete {
+		secretRef := secretsToDelete[i]
+		serviceAccount, err := findServiceAccountForSecret(ctx, secretRef, serviceAccounts)
+		if err != nil {
+			deletionErr = errors.Join(deletionErr, err)
+			continue
+		}
+
+		// If have a ServiceAccount for this secret, then we can leave it in
```

Comment:
typo: missing `we`

Also you can probably hoist the orphaned `place.` to the end of this line (nanoquibbling...)